### ISSUE TITLE
udp connback doesn't work unless accepter is also udp

### ISSUE THREAD

**Issue by andrewwade** [cminyard/ser2net#50]
Using version 4.3.3.

I tried making ser2net act as a client only, which to me, means I don't want an accepter. I set `accepter` to `null` and set `connback` to `tcp,localhost,1234` which worked fine. 

Then I switched `connback` to `udp,localhost,1234` and I got this error:

`ser2net[7903]: Unable to allocate connect back port server, addr udp,localhost,1234: Operation not supported`

Then I switched the `accepter` to be `udp,8888` and it now worked as expected. But now ser2net is acting as a server on port 8888 which I don't want.

**Comment by cminyard** [comment]
On Tue, May 25, 2021 at 09:28:37PM -0700, Andrew Wade wrote:
> Using version 4.3.3.
> 
> I tried making ser2net act as a client only, which to me, means I don't want an accepter. I set `accepter` to `null` and set `connback` to `tcp,localhost,1234` which worked fine. 
> 
> Then I switched `connback` to `udp,localhost,1234` and I got this error:
> 
> `ser2net[7903]: Unable to allocate connect back port server, addr udp,localhost,1234: Operation not supported`
> 
> Then I switched the `accepter` to be `udp,8888` and it now worked as expected. But now ser2net is acting as a server on port 8888 which I don't want.

Yes, that is the expected behaviour, documented in the man page.  If you
do not increase the number of connections allowed, the server port
will not accept an incoming connection, but it will have the port open.
For UDP you aren't really wasting anything, you are just setting the
source port for the UDP connection.  In reality it is really no
different than if you disabled the accepter.  You have to have a source
port for UDP.  For TCP it's different, of course.  If you want a random
source port, you can set the port to 0.

To do what you want, you can use the conacc gensio, which converts a
connector gensio to an accepter.  For the accepter, you can do:

  conacc,udp,localhost,1234

and not use connback.  You probably want to disable the timeout in this
scenario.  I don't think I have tried this particular scenario, but it
should work.  But in practical terms it is no different than using
connback.

-corey

**Comment by andrewwade** [comment]
Thanks for the quick reply. I'll try your suggestion tonight. I have a few follow-up questions:

1. How is it different for TCP? Doesn't TCP also require a source port?

2. If I use your suggestion `conacc,udp,localhost,1234`, what source port does it use?

3. What if I want to have a client that connects to multiple servers (a.k.a multiple connectors)? Would I need to use the connback parameter in that case which means I'd have to set the accepter to `udp,<source port>` and then set `max-connections` to the number of connback connections?

4. While I have you, do you have a multicast example where serial can send and receive on a multicast group? I'm just getting familiar with the gensio library and I see some multicast options, but I haven't tried them yet.

**Comment by cminyard** [comment]
On Wed, May 26, 2021 at 03:59:02PM -0700, Andrew Wade wrote:
> Thanks for the quick reply. I'll try your suggestion tonight. I have a few follow-up questions:
> 
> 1. How is it different for TCP? Doesn't TCP also require a source port?

Yes, but the source port in the accepter line does not set the TCP
source port.  That's picked randomly by the kernel unless you override
it with the laddr= option.  It's inconsistent, but that might cause
problems with TCP that you couldn't work around.

> 
> 2. If I use your suggestion `conacc,udp,localhost,1234`, what source port does it use?

It randomly picks one.  There is a laddr=<addr> option to udp that lets
you set the source port, documented in gensio(5).

> 
> 3. What if I want to have a client that connects to multiple servers (a.k.a multiple connectors)? Would I need to use the connback parameter in that case which means I'd have to set the accepter to `udp,<source port>` and then set `max-connections` to the number of connback connections?
> 

In that case you would need to use connback, it won't work with conacc.
And you are correct, set the max-connections to the number of connbacks.

> 4. While I have you, do you have a multicast example where serial can send and receive on a multicast group? I'm just getting familiar with the gensio library and I see some multicast options, but I haven't tried them yet.

I don't have an example for that besides what's in the udp section of
gensio(5).  If you want to send to a multicast address, I believe you
can just set the destination address to a multicast address and it will
send to the given address.  It's more complicated for receiving
multicast packets, but that's probably not as useful.

-corey

**Comment by andrewwade** [comment]
Thanks for answering my questions. I'll mess around with multicast when I get a chance. I think being able to receive from multicast would be useful in my application. But I think I can probably use socat to forward multicast traffic to ser2net if necessary.

One last question:

Why do all connback's have to connect before sending data to any of them? Is there a way to change it so that data start sending as soon as the first connback is connected then as another connback connects later, it automatically starts sending to that one as well? 

Basically I have a network where connbacks can come in and out of the network at random so I'd like to provide a list of connbacks and I'd like ser2net to communicate with whichever conbacks are connected at any given time.

**Comment by cminyard** [comment]
On Thu, May 27, 2021 at 03:42:45PM -0700, Andrew Wade wrote:
> Thanks for answering my questions. I'll mess around with multicast when I get a chance. I think being able to receive from multicast would be useful in my application. But I think I can probably use socat to forward multicast traffic to ser2net if necessary.
> 
> One last question:
> 
> Why do all connback's have to connect before sending data to any of them? Is there a way to change it so that data start sending as soon as the first connback is connected then as another connback connects later, it automatically starts sending to that one as well? 

It complicates things more than you might imagine.  When you have
multiple connections, flow-control becomes a hard issue.  Do you delay
every connection because one is slow?  Or do you drop data?  ser2net
will delay every connection as it's policy, but then unconnected ones
become an issue.  I could make it where if nothing is connected, the
serial port is flow-controlled, but if something is connected it is not
flow-controlled and the data flows to existing connections.  It would
take some care, though.  And others will get data dropped.

Multiple connections is kind of a hack no matter what you do.

> 
> Basically I have a network where connbacks can come in and out of the network at random so I'd like to provide a list of connbacks and I'd like ser2net to communicate with whichever conbacks are connected at any given time.

I thought you were using UDP?  If that's the case, it should work fine,
since ser2net doesn't know that nothing's on the other end and just
sends into the oblivian.  Multicast would be even better.  If it's TCP, 
then that's an issue.

-corey
